docs/pallas/tpu/distributed.md

```diff
@@ -97,10 +97,23 @@ The `pltpu.make_async_remote_copy` function is used to create a remote DMA descr
 dma_descriptor = make_async_remote_copy(src_ref, dst_ref, send_sem, recv_sem, device_id)
 dma_descriptor.start() # Initiate the DMA (non-blocking).
 # ... do other work
-dma_descriptor.wait_send() # Block until all data has been sent.
-dma_descriptor.wait_recv() # Block until all data has been received.
+
+# Block until all data has been sent from this device.
+# Now it's safe to overwrite the data in src_ref on this device.
+dma_descriptor.wait_send()
+
+# Block until all data has been received by this device.
+# Now it's safe to start reading from dst_ref on this device.
+# This will hang unless somebody makes an identically shaped DMA with
+# device_id set to this device.
+dma_descriptor.wait_recv()
+
+# How do you know when it's safe to make another DMA to the same
+# dst_ref location on `device_id`? Read on..
 ```
 
+Note that `dma_descriptor.wait_recv()` does not mean waiting for `device_id` to receive all the data. Instead it means waiting for some other device to write to `dst_ref` on the _current device_ and signal its `recv_sem` by the expected value.
```

Comment:
I think there's some nuance with how the API is set up since it assumes a "SIMD" mode of operation but you can circumvent that with branching. Maybe something like the following:

Note that in this API, all Refs and semaphores refer to those that are locally allocated on the current device. It is assumed that the `dst_ref` and `recv_sem` are also allocated on the destination `device_id` by the time the DMA executes (which is guaranteed by proper usage of barrier semaphores, which are by default automatically managed by Pallas but can also be used manually). The semantics of `wait_recv()` are for those of the local (receiving) device and means that it waits for another device to write to the `dst_ref` allocated locally and signal `recv_sem`.